File: src/test/java/pl/training/workshop/m6/s11_safecomposite/S11SolutionTest.java
package pl.training.workshop.m6.s11_safecomposite;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import org.junit.jupiter.api.Test;

/** Transparent kontra Safe: gdzie wychodzi błąd add() na liściu. */
final class S11SolutionTest {
    /** Dokumentuje pułapkę start; po "warsztat.sh jump" (start = krok 1+) test jest pomijany. */
    @Test
    void transparentCompositeFailsAtRuntime() throws ReflectiveOperationException {
        String start = "pl.training.workshop.m6.s11_safecomposite.start.";
        Class<?> component = Class.forName(start + "MenuComponent");
        Class<?> product = Class.forName(start + "Product");
        Method add = findMethod(product, "add", component);
        assumeTrue(add != null, "start nie jest już Transparent Composite");
        Object nachos = product.getConstructor(String.class, String.class).newInstance("Nachos", "14.00");
        Object sauce = product.getConstructor(String.class, String.class).newInstance("Sos", "3.00");
        var error = assertThrows(InvocationTargetException.class, () -> add.invoke(nachos, sauce));
        assertEquals(UnsupportedOperationException.class, error.getCause().getClass());
        assertEquals("cannot add to Nachos", error.getCause().getMessage());
    }

    private static Method findMethod(Class<?> type, String name, Class<?>... parameters) {
        try {
            return type.getMethod(name, parameters);
        } catch (NoSuchMethodException exception) {
            return null;
        }
    }

    @Test
    void safeCompositeHasNoAddOnLeafOrCommonType() {
        String step1 = "pl.training.workshop.m6.s11_safecomposite.step1.";
        assertThrows(NoSuchMethodException.class, () -> Class.forName(step1 + "Product")
                .getMethod("add", Class.forName(step1 + "MenuComponent")));
        assertThrows(NoSuchMethodException.class, () -> Class.forName(step1 + "MenuComponent")
                .getMethod("children"));
    }

    @Test
    void immutableCompositeHasNoAddAtAll() {
        var combo = pl.training.workshop.m6.s11_safecomposite.step2.Combo.of("Zestaw");
        assertThrows(UnsupportedOperationException.class, () -> combo.children().add(null));
    }
}

File: csharp/tests/Training.Workshop.Tests/M6/S11SafeComposite/S11SolutionTest.cs
using System.Reflection;
using Training.Workshop.Shared;

namespace Training.Workshop.Tests.M6.S11SafeComposite;

/// <summary>Transparent kontra Safe: gdzie wychodzi błąd Add() na liściu.</summary>
public sealed class S11SolutionTest
{
    private const string Scene = "Training.Workshop.M6.S11SafeComposite.";

    /// <summary>
    /// Dokumentuje pułapkę Start; po "warsztat.sh jump" (Start = krok 1+) test kończy się
    /// bez sprawdzeń (xUnit 2 nie ma odpowiednika assumeTrue).
    /// </summary>
    [Fact]
    public void TransparentCompositeFailsAtRuntime()
    {
        var component = TypeOf(Scene + "Start.MenuComponent");
        var product = TypeOf(Scene + "Start.Product");
        var add = product.GetMethod("Add", [component]);
        if (add == null)
        {
            return; // Start nie jest już Transparent Composite
        }
        var nachos = Activator.CreateInstance(product, "Nachos", "14.00");
        var sauce = Activator.CreateInstance(product, "Sos", "3.00");
        var error = Assert.Throws<TargetInvocationException>(() => add.Invoke(nachos, [sauce]));
        Assert.Equal(typeof(NotSupportedException), error.InnerException!.GetType());
        Assert.Equal("cannot add to Nachos", error.InnerException.Message);
    }

    [Fact]
    public void SafeCompositeHasNoAddOnLeafOrCommonType()
    {
        var component = TypeOf(Scene + "Step1.MenuComponent");
        Assert.Null(TypeOf(Scene + "Step1.Product").GetMethod("Add", [component]));
        Assert.Null(component.GetProperty("Children"));
    }

    [Fact]
    public void ImmutableCompositeHasNoAddAtAll()
    {
        var combo = Training.Workshop.M6.S11SafeComposite.Step2.Combo.Of("Zestaw");
        var children = (ICollection<Training.Workshop.M6.S11SafeComposite.Step2.MenuComponent>)combo.Children;
        Assert.Throws<NotSupportedException>(() => children.Add(null!));
    }

    private static Type TypeOf(string name) =>
        typeof(Money).Assembly.GetType(name) ?? throw new TypeLoadException(name);
}
